src/java/org/apache/cassandra/index/sai/utils/IndexTermType.java

```diff
@@ -0,0 +1,858 @@
+    {
+        this.columnMetadata = columnMetadata;
+        this.indexTargetType = indexTargetType;
+        this.capabilityBitmap = calculateIdentityBitmap(columnMetadata, partitionColumns, indexTargetType);
+        this.indexType = calculateIndexType(columnMetadata.type, capabilityBitmap, indexTargetType);
+        if (indexType.subTypes().isEmpty())
+        {
+            this.subTypes = Collections.emptyList();
+        }
+        else
+        {
+            List<IndexTermType> subTypes = new ArrayList<>(indexType.subTypes().size());
+            for (AbstractType<?> subType : indexType.subTypes())
+                subTypes.add(new IndexTermType(columnMetadata.withNewType(subType), partitionColumns, indexTargetType));
+            this.subTypes = Collections.unmodifiableList(subTypes);
+        }
+        if (isVector())
+        {
+            VectorType<?> vectorType = (VectorType<?>) indexType;
+            vectorElementType = vectorType.elementType;
+            vectorDimension = vectorType.dimension;
+        }
+        else
+        {
+            vectorElementType = null;
+            vectorDimension = -1;
+        }
+    }
+
+    /**
+     * Returns {@code true} if the index type is a literal type and will use a literal index. This applies to
+     * string types, frozen types, composite types and boolean type.
+     */
+    public boolean isLiteral()
+    {
+        return hasProperty(IS_LITERAL);
+    }
+
+    /**
+     * Returns {@code true} if the index type is a string type. This is used to determine if the type supports
+     * analysis.
+     */
+    public boolean isString()
+    {
+        return hasProperty(IS_STRING);
+    }
+
+    /**
+     * Returns {@code true} if the index type is a vector type. Note: being a vector type does not mean that the type
+     * is valid for indexing in that we don't check the element type and dimension constraints here.
+     */
+    public boolean isVector()
+    {
+        return hasProperty(IS_VECTOR);
+    }
+
+    /**
+     * Returns {@code true} if the index type is reversed. This is only the case (currently) for clustering keys with
+     * descending ordering.
+     */
+    public boolean isReversed()
+    {
+        return hasProperty(IS_REVERSED);
+    }
+
+    /**
+     * Returns {@code true} if the index type is frozen, e.g. the type is wrapped with {@code frozen<type>}.
+     */
+    public boolean isFrozen()
+    {
+        return hasProperty(IS_FROZEN);
+    }
+
+    /**
+     * Returns {@code true} if the index type is a non-frozen collection
+     */
+    public boolean isNonFrozenCollection()
+    {
+        return hasProperty(IS_NON_FROZEN_COLLECTION);
+    }
+
+    /**
+     * Returns {@code true} if the index type is a frozen collection. This is the inverse of a non-frozen collection
+     * but this method is here for clarity.
+     */
+    public boolean isFrozenCollection()
+    {
+        return !hasProperty(IS_NON_FROZEN_COLLECTION);
```

Comment:
If the type isn't a non-frozen collection, that doesn't mean it _is_ a frozen collection, or a collection at all, right? I think this is part of the problem in `IndexTermTypeTest`. Tuples and UDTs are totally not non-frozen collections, but it's because they're not collections.

Reply:
Correct, I've fixed on [62970c5](https://github.com/apache/cassandra/pull/2916/commits/62970c5d28560d20cef503bf05299a5a9e6a6c1d). I was trying to optimise the method and completely forgot about tuples and UDTs.